### PR TITLE
Finality notification streams

### PR DESCRIPTION
**Pull request by rphmeier**
Almost done with #746 : now we can actually finalize blocks in the client and get a notification stream for them.

Also adds tests for `tree_route`, having moved it to common client code, and doing implicit ancestry finalization in the client as opposed to at the DB layer (there is room for optimization here but I'll leave it for now as in general we'll finalize only a few blocks at a time).

### CODE REVIEW

File: core/client/src/client.rs [hunk 1]
```diff
@@ -436,6 +455,19 @@ impl<B, E, Block> Client<B, E, Block> where
 			blockchain::BlockStatus::Unknown => {},
 		}
 
+		let last_best = self.backend.blockchain().info()?.best_hash;
+
+		let make_notifications = match origin {
+			BlockOrigin::NetworkBroadcast | BlockOrigin::Own | BlockOrigin::ConsensusBroadcast => true,
+			_ => false,
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply core/client/src/client.rs]: @arkpar should give this justification -- I'm leaving the logic the same as it was before and I don't know exactly why it was chosen this way.

[the remark this answers is not in the document]
Comment (rphmeier) [reply core/client/src/client.rs]: IMO `BlockOrigin` is an approximation for a bool `make_notifications` which should be the real parameter and determined by the sync service when scheduling the import.

**Comment by arkpar** [comment]
Looks like this affects sync test somehow.

**Comment by pepyakin** [comment]
Yeah, I canceled the job https://gitlab.parity.io/parity/substrate/-/jobs/100459 because it running for 152 minutes already.

Relevant log:

```
Running target/release/deps/substrate_network-e75ef8dd93caa92c
...
test test::sync::sync_from_two_peers_with_ancestry_search_works ... test test::sync::sync_from_two_peers_with_ancestry_search_works has been running for over 60 seconds
test test::sync::sync_after_fork_works ... test test::sync::sync_after_fork_works has been running for over 60 seconds
test test::sync::sync_no_common_longer_chain_fails ... test test::sync::sync_no_common_longer_chain_fails has been running for over 60 seconds
```

**Comment by rphmeier** [comment]
The offending sync tests were those that required blocks to be reorganized from one fork, and then spinning in a loop while the client refused to revert unfinalized blocks.

**Pull request merged**